### ISSUE TITLE
About the cost of GPU during test

### ISSUE THREAD

**Issue by lyc6749** [facebookresearch/detectron2#321]
## ❓ Questions and Help

when I use the trained maskrcnn-resnet50-FPN model  for testing， I found it cost about 7~8G GPU memory which is more than the version of maskrcnn-benchmark. How can I do to decrease the 
consumption of GPU?

**Comment by ppwwyyxx** [comment]
```
python tools/train_net.py --config-file configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_1x.yaml --num-gpus 1 --eval-only
```
only takes 2G memory on my machine. If this is the case for you as well, then it means your code is not written correctly.
It's unlikely, but if the same commands takes 8G memory on your environment, then it's an environment issue and please include environment details following the issue template.

**Comment by lyc6749** [comment]
# packages in environment at /home/chenlingying/anaconda3/envs/detectron2:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main
absl-py                   0.8.1                    pypi_0    pypi
addict                    2.2.1                    pypi_0    pypi
blas                      1.0                         mkl
ca-certificates           2019.10.16                    0
cachetools                3.1.1                    pypi_0    pypi
certifi                   2019.9.11                py36_0
cffi                      1.13.1           py36h2e261b9_0
chardet                   3.0.4                    pypi_0    pypi
cloudpickle               1.2.2                    pypi_0    pypi
cudatoolkit               10.1.243             h6bb024c_0
cycler                    0.10.0                   pypi_0    pypi
cython                    0.29.14                  pypi_0    pypi
decorator                 4.4.1                    pypi_0    pypi
detectron2                0.1                       dev_0    <develop>
freetype                  2.9.1                h8a8886c_1
fvcore                    0.1                      pypi_0    pypi
google-auth               1.6.3                    pypi_0    pypi
google-auth-oauthlib      0.4.1                    pypi_0    pypi
grpcio                    1.25.0rc1                pypi_0    pypi
idna                      2.8                      pypi_0    pypi
imagecorruptions          1.1.0                    pypi_0    pypi
imageio                   2.6.1                    pypi_0    pypi
imgaug                    0.2.6                    pypi_0    pypi
intel-openmp              2019.4                      243
jpeg                      9b                   h024ee3a_2
kiwisolver                1.1.0                    pypi_0    pypi
libedit                   3.1.20181209         hc058e9b_0
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 9.1.0                hdf63c60_0
libgfortran-ng            7.3.0                hdf63c60_0
libpng                    1.6.37               hbc83047_0
libstdcxx-ng              9.1.0                hdf63c60_0
libtiff                   4.0.10               h2733197_2
markdown                  3.1.1                    pypi_0    pypi
matplotlib                3.2.0rc1                 pypi_0    pypi
mkl                       2019.4                      243
mkl-service               2.3.0            py36he904b0f_0
mkl_fft                   1.0.14           py36ha843d7b_0
mkl_random                1.1.0            py36hd6b4f25_0
mmcv                      0.2.14                   pypi_0    pypi
mmdet                     1.0rc0+4b101ec            dev_0    <develop>
ncurses                   6.1                  he6710b0_1
networkx                  2.4                      pypi_0    pypi
ninja                     1.9.0            py36hfd86e86_0
numpy                     1.17.2           py36haad9e8e_0
numpy-base                1.17.2           py36hde5b4d6_0
oauthlib                  3.1.0                    pypi_0    pypi
olefile                   0.46                     py36_0
opencv-python             4.1.1.26                 pypi_0    pypi
openssl                   1.1.1d               h7b6447c_3
pillow                    6.2.0            py36h34e0f95_0
pip                       19.3.1                   py36_0
portalocker               1.5.1                    pypi_0    pypi
protobuf                  3.10.0                   pypi_0    pypi
pyasn1                    0.4.7                    pypi_0    pypi
pyasn1-modules            0.2.7                    pypi_0    pypi
pycocotools               2.0                      pypi_0    pypi
pycparser                 2.19                     py36_0
pyparsing                 2.4.2                    pypi_0    pypi
python                    3.6.9                h265db76_0
python-dateutil           2.8.1                    pypi_0    pypi
pytorch                   1.3.0           py3.6_cuda10.1.243_cudnn7.6.3_0    pytorch
pywavelets                1.1.1                    pypi_0    pypi
pyyaml                    5.1.2                    pypi_0    pypi
readline                  7.0                  h7b6447c_5
requests                  2.22.0                   pypi_0    pypi
requests-oauthlib         1.2.0                    pypi_0    pypi
rsa                       4.0                      pypi_0    pypi
scikit-image              0.16.2                   pypi_0    pypi
scipy                     1.3.1                    pypi_0    pypi
setuptools                41.6.0                   py36_0
six                       1.12.0                   py36_0
sqlite                    3.30.1               h7b6447c_0
tensorboard               2.0.1                    pypi_0    pypi
termcolor                 1.1.0                    pypi_0    pypi
tk                        8.6.8                hbc83047_0
torchvision               0.4.1                py36_cu101    pytorch
tqdm                      4.37.0                   pypi_0    pypi
urllib3                   1.25.6                   pypi_0    pypi
werkzeug                  0.16.0                   pypi_0    pypi
wheel                     0.33.6                   py36_0
xz                        5.2.4                h14c3975_4
zlib                      1.2.11               h7b6447c_3
zstd                      1.3.7                h0b5b093_0
This is my environment details. I also changed the num_workers to 1 , but nothing chnagged.

**Comment by ppwwyyxx** [comment]
Please also include the command you run, the full logs you observed, and how you measure GPU memory usage.

**Comment by lyc6749** [comment]
$ python -m detectron2.utils.collect_env
------------------------  -------------------------------------------------------------------
sys.platform              linux
Python                    3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 19:07:31) [GCC 7.3.0]
Numpy                     1.17.2
Detectron2 Compiler       GCC 5.4
Detectron2 CUDA Compiler  10.1
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.3.0
PyTorch Debug Build       False
torchvision               0.4.1a0+d94043a
CUDA available            True
GPU 0,1,2,3               TITAN X (Pascal)
CUDA_HOME                 /home/xxx/cuda-10.1
NVCC                      Cuda compilation tools, release 10.1, V10.1.243
Pillow                    6.2.0
cv2                       4.1.1
------------------------  -------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_50,code=compute_50
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

**Comment by lyc6749** [comment]
I measure the usage of GPU just by 
![image](https://user-images.githubusercontent.com/33708958/68918326-a2799a80-07a8-11ea-803c-9aab63e32b83.png)

**Comment by ppwwyyxx** [comment]
> Please also include the command you run, the full logs you observed following the issue template

Could you provide this?

Also, in case you did not see, this is what I commented above:
```
python tools/train_net.py --config-file configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_1x.yaml --num-gpus 1 --eval-only
```
only takes 2G memory on my machine. If this is the case for you as well, then it means your inference code is not written correctly.

**Comment by lyc6749** [comment]
I run the command like this :
````
CUDA_VISIBLE_DEVICES=3 python tools/train_net.py --config-file configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_1x.yaml --num-gpus 1 --eval-only MODEL.WEIGHTS /data/xxx/xxx/detectron2/output/model_final.pth
```

**Comment by ppwwyyxx** [comment]
I ran the command 
```
python tools/train_net.py --config-file configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_1x.yaml --num-gpus 1 --eval-only MODEL.WEIGHTS detectron2://COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_1x/137260431/model_final_a54504.pkl
```
with no modifications to code and on two machines I both observed <3G memory consumption at the end of the command. It seems I'm unable to reproduce your observations and cannot help much with your situation. Some more reproducible instructions (e.g. a docker/colab) may help.

**Comment by bendidi** [comment]
i've also had the same issue, during test on GTX 1080, the model uses 7Gb of memory with a batch size of 1 (maybe caused by loading the optimizer and scheduler as well ?) the training was done with a batch size of 4 on a V100 Gpu

**Comment by ppwwyyxx** [comment]
Closing as there are no instructions to reproduce the issue.

**Comment by bendidi** [comment]
I can maybe share a `model.pth` and `config.yaml` @ppwwyyxx to reproduce?

**Comment by ppwwyyxx** [comment]
It's easier for you to run the public model using the command above to see whether you still experience memory issues. If not, then it would make sense to look at your models to see what makes it less efficient.

**Comment by bendidi** [comment]
I've shared a `model.pth` and `config.yaml` here @ppwwyyxx : https://drive.google.com/drive/folders/1LUcq2DwNowxJIH59yVHJo5zSJx6kaYY9?usp=sharing

**Comment by ppwwyyxx** [comment]
Your config does not contain datasets so it cannot be run. As I said it would be easier if we use the public model and the official config.

Running your config & model on coco_2017_val costs 3.4G memory on GTX1080 inside the official detectron2 docker.

**Comment by bendidi** [comment]
I've created a repo for reproducing this: https://github.com/bendidi/test-gpu-cost-d2 (same model as before, direct link in the README.md)

Test environment: EC2 instance (p2.xlarge)

```
[12/22 18:45:49 detectron2]: Rank of current process: 0. World size: 1
[12/22 18:45:49 detectron2]: Environment info:
------------------------  --------------------------------------------------
sys.platform              linux
Python                    3.7.5 (default, Oct 25 2019, 15:51:11) [GCC 7.3.0]
Numpy                     1.18.0
Detectron2 Compiler       GCC 5.4
Detectron2 CUDA Compiler  10.0
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.3.1
PyTorch Debug Build       False
torchvision               0.4.2
CUDA available            True
GPU 0                     Tesla K80
CUDA_HOME                 /usr/local/cuda
NVCC                      Cuda compilation tools, release 10.0, V10.0.130
Pillow                    6.2.1
cv2                       4.1.2
------------------------  --------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,
```

nvidia-smi: 

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 418.67       Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla K80           On   | 00000000:00:1E.0 Off |                    0 |
| N/A   66C    P0   125W / 149W |   6414MiB / 11441MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      3839      C   python                                      6403MiB |
+-----------------------------------------------------------------------------+
```

inference speed: 

```
[12/22 18:46:24 detectron2]: Num_instances=24 => Time: 0.9872090816497803 seconds.
[12/22 18:46:25 detectron2]: Num_instances=24 => Time: 0.9864170551300049 seconds.
[12/22 18:46:26 detectron2]: Num_instances=24 => Time: 0.9957678318023682 seconds.
[12/22 18:46:27 detectron2]: Num_instances=24 => Time: 0.9992132186889648 seconds.
[12/22 18:46:28 detectron2]: Num_instances=24 => Time: 1.000138759613037 seconds.
[12/22 18:46:29 detectron2]: Num_instances=24 => Time: 0.996544599533081 seconds.
[12/22 18:46:30 detectron2]: Num_instances=24 => Time: 0.9902329444885254 seconds.
[12/22 18:46:31 detectron2]: Num_instances=24 => Time: 0.9826292991638184 seconds.
[12/22 18:46:32 detectron2]: Num_instances=24 => Time: 0.9849972724914551 seconds.
```

**Comment by ppwwyyxx** [comment]
Your test image has a higher resolution and simply representing the masks of 24 results will take a few GBs of memory. This is by definition of the task itself.

Nothing unexpected is observed.

Re. benchmarking, using `DefaultPredictor` will include the overhead to resize an image into a target size using the slow PIL library. So we do not use it to do any benchmarking.
